### PR TITLE
Push complement image to a docker registry

### PR DESCRIPTION
**Pull request by michaelkaye**
This PR automates the building of complement images for key branches (main/master, develop). A few projects eg matrix-org/trafficlight & [matrix-appservice-irc](https://github.com/matrix-org/matrix-appservice-irc/blob/develop/.github/workflows/e2e-test.yml#L11) are using `homerunner` from complement to manage setup and teardown of synapse and dendrite servers within tests.

We're noticing that we're asking github actions or developers to locally create these images each time we run our tests - but actually that's not relevant for using homerunner to start integration tests of clients etc. Having a registry would let us avoid running out of date synapse servers and hopefully speed up test times on GHA.

I chose (for now at least) to keep this workflow separate from the docker.yaml  - We login to a different repository and probably don't want to share secrets with this build process (the complement image has a specific script to build). **Should we combine these workflows?**, or we could even incorporate the build process directly into the GHA - which might enable us to keep the same labels as are used on the main synapse image.

I chose not to publish the workers docker image as it's an intermediate one that doesn't seem to have much use outside of locally built developer machines. **Should we publish this?**.

I chose to push to ghcr.io rather than docker-hub to reduce the visibility of the created image - we don't want people using these images for anything other than testing using complement; and this is in many ways just a cache of the build. Before we merge the PR, we should probably **consider changing the location or naming of these images**

### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [X] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [X] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by michaelkaye** [comment]
https://github.com/michaelkaye/synapse/actions/runs/3516474915 is the build running on my fork of synapse.

File: .github/workflows/complement.yml
Comment (DMRobertson) [inline .github/workflows/complement.yml:7]: Note that Synapse's primary branch is still called `master`

Comment (michaelkaye) [reply .github/workflows/complement.yml]: Ah, yes; i was cribbing from another repo. Thanks for the catch.

File: .github/workflows/complement.yml
Comment (reivilibre) [inline .github/workflows/complement.yml:1]: would be tempted to rename the file so it's more obvious that this is only for uploading the complement image.
`push_complement_image.yml` would be fine IMO.

File: .github/workflows/complement.yml
Comment (reivilibre) [inline .github/workflows/complement.yml:7]: this is basically entirely up to you, but do you want to push images for tagged releases?
I could see it maybe being useful for comparing different releases if necessary.

**Review by reivilibre — CHANGES_REQUESTED** [review]
LGTM otherwise

File: .github/workflows/complement.yml
Comment (michaelkaye) [reply .github/workflows/complement.yml:1]: renamed per suggestion.

File: .github/workflows/complement.yml
Comment (michaelkaye) [reply .github/workflows/complement.yml:7]: For my test cases, we probably only want to test current and next as well; if people want to investigate certain synapse versions then that can happen as a local build for now.

**Comment by michaelkaye** [comment]
It looks like this is OK from a technical standpoint - would like to confirm that my questions in bold have been considered before we merge and start making these artifacts public.

File: .github/workflows/push_complement_image.yml
Comment (DMRobertson) [inline .github/workflows/push_complement_image.yml:7]: Can you add per-branch concurrency limits here, do limit GHA worker usage when merge a bunch of PRs in short order? See e.g. https://github.com/matrix-org/sytest/blob/bebf431edf7311d5147139f44dab97131a784d29/.github/workflows/pipeline.yml#L7-L12 from sytest

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Discussed during triage meeting. @erikjohnston mentioned that doing this for every `develop` build might eat up into our pool of github actions workers. We might consider only running this nightly for develop in the future, but this is probably fine for now.

> Should we combine these workflows?

Let's keep them separate.

> Should we publish [the workers docker image]?

I think no (for now). The with-workers docker image is fairly new (and was only created recently, to test Synapse with workers under complement). I think it sees some use by the community but not a whole lot.

More generally: I think it's the Synapse team's job to make sure that Synapse-polylith behaves like Synapse-polylith; I don't think it's on downstream testers to cover that use case.

> Before we merge the PR, we should probably consider changing the location or naming of these images

`complement-synapse` is probably fine (though the name "complement"  doesn't obviously mean "test purposes only do not use in production"). As long as we're not advertising the existence of this image, I think it's fine.

**Comment by michaelkaye** [comment]
I've made the appropriate changes (i believe) to run the build at 5am on develop and on push to master, as well as cancelling builds that are in flight. Testing this is always a bit weird so would benefit from someone with write access to repo doing the merge and watching it.

**Review by DMRobertson — APPROVED** [review]
I'll fixup the crontab; otherwise LGTM. Let's get it merged and try it out.

(I might have to ask you to check back tomorrow to see if this build and pushed a sensible image overnight)

**Comment by DMRobertson** [comment]
Hmm. Why is https://github.com/matrix-org/synapse/actions/runs/3564702663 unhappy?

**Comment by DMRobertson** [comment]
> Hmm. Why is [matrix-org/synapse/actions/runs/3564702663](https://github.com/matrix-org/synapse/actions/runs/3564702663) unhappy?

Guessing it didn't like the yaml syntax. I've tried to fix this up. I've also stuck in something to try and let us run this once it's merged.

**Comment by DMRobertson** [comment]
> The workflow is not valid. .github/workflows/push_complement_image.yml (Line: 37, Col: 13): Unexpected symbol: '"workflow_dispatch"'. Located at position 22 within expression: github.event.type == "workflow_dispatch" .github/workflows/push_complement_image.yml (Line: 42, Col: 13): Unexpected symbol: '"schedule"'. Located at position 22 within expression: github.event.type == "schedule"

I hate GHA sometimes. Maybe we should use https://github.com/rhysd/actionlint in the future.

File: .github/workflows/push_complement_image.yml
Comment (DMRobertson) [inline .github/workflows/push_complement_image.yml:66]: I wonder if `master` should point to latest and `develop` should have a different tag?

**Comment by michaelkaye** [comment]
> > The workflow is not valid. .github/workflows/push_complement_image.yml (Line: 37, Col: 13): Unexpected symbol: '"workflow_dispatch"'. Located at position 22 within expression: github.event.type == "workflow_dispatch" .github/workflows/push_complement_image.yml (Line: 42, Col: 13): Unexpected symbol: '"schedule"'. Located at position 22 within expression: github.event.type == "schedule"
> 
> I hate GHA sometimes. Maybe we should use https://github.com/rhysd/actionlint in the future.

Oh wow, this is what i've wanted in my life for so long.

**Pull request merged**
